### PR TITLE
sql: introduce tenant RO setting to allow full access to zone configs

### LINKED ISSUE #75569: multi-tenant: Relax restrictions on supported zone configuration changes from secondary tenants
Originally we were of the opinion that zone configurations in secondary tenants would only be updated by the multi-region semantics.  As a result, we restricted their use to only specify constraints which referenced the region or zone tiers:

https://github.com/cockroachdb/cockroach/blob/81c447ddb47e33d66f59acbcb1f76c37f5a148f6/pkg/sql/set_zone_config.go#L991-L1020 

The potential problem with this approach is that if we're going to make multi-tenant deployments the default deployment pattern for CRDB (something which is still being debated), this restriction will break some users.  We'll need to relax this restriction if/when we go down that route.

Jira issue: CRDB-12730

Epic CRDB-26686

### LINKED ISSUE #111358: sql: introduce a cluster setting to by-pass zone config restrictions in virtual clusters
Until #100787 is implemented, we have disabled the ability to set some zone configurations in virtual clusters. For UA we want to enable operators to allow virtual clusters to by-pass that. As the quick solution we'll add a cluster setting for it.

Jira issue: CRDB-31871
Epic: CRDB-26686

### PR DESCRIPTION
**Pull request by yuzefovich**
TLDR: this commit introduces another cluster setting that controls
whether the virtual clusters have unrestricted ability to modify zone
configs.

Previously, we already had `sql.zone_configs.allow_for_secondary_tenant.enabled`
setting (which controls whether VCs can modify zone configs at all), but
even if that was enabled, VCs could not modify constraints and lease
preferences other than locality (and only regions and zones). However,
this restriction can be too cumbersome in UA, so this commit introduces
another cluster setting
`sql.virtual_cluster.feature_access.zone_configs_unrestricted.enabled`
that controls whether VCs have full access to zone config modification
(note that region and zone validation is still preserved, and retrieval
of that information is made lazy). The new setting is disabled by
default, but it's overridden in the config profiles `virtual+noapp`,
`virtual+noapp+repl`, `virtual+app+sharedservice`, and
`virtual+app+sharedservice+repl`.

Fixes: #75569.
Fixes: #111358.
Epic: CRDB-26686

Release note: None

### CODE REVIEW

**Review by knz — APPROVED** [review]
:lgtm: with nits

> Reviewed 7 of 9 files at r2, 15 of 15 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111387#-:-NfQviezACmwizyzFJI-:bfmeux1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @stevendanna and @yuzefovich)

___
*[`-- commits` line 5 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111387#-NfQun9P-_wxJd75bckT):*
nit: since the first sentence is a TLDR, place it on a line of its own, with a paragraph break with what follows.
___
*[`-- commits` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111387#-NfQutu76cXIfaAPnGfU):*
pls include the name of that setting in the commit message.
___
*[`-- commits` line 15 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111387#-NfQuxN4AZTynUl079Kj):*
pls include the name of which profile(s) was updated.
___
*[`pkg/sql/set_zone_config.go` line 1083 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111387#-NfQv5-Q83YLtIMRU6aK:-NfQv5-Q83YLtIMRU6aL:b2n1wh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a4ab3c1091972fb38172b860424be499191d819/pkg/sql/set_zone_config.go#L1083)):*
> ```Go
> 	settings.TenantReadOnly,
> 	"sql.virtual_cluster.feature_access.zone_configs_all.enabled",
> 	"enable full usage of ALTER CONFIGURE ZONE in virtual clusters",
> ```

nit: "full" -> "unrestricted"

**Review by yuzefovich — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/111387#-:-NfRbqRg2scWOp4sLrty:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @knz and @stevendanna)

___
*[`-- commits` line 11 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111387#-NfQutu76cXIfaAPnGfU):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

pls include the name of that setting in the commit message.
</blockquote></details>

Done.
___
*[`-- commits` line 15 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/111387#-NfQuxN4AZTynUl079Kj):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

pls include the name of which profile(s) was updated.
</blockquote></details>

Done.
